WIP interactive visualization for volumetric stcs

jasmainak opened the pull request:
The goal is to create a Nutmeg style GUI [like this](https://www.youtube.com/watch?v=r3tGuZJHzT4)

cc @sarangnemo @agramfort @k-jerbi

I got something basic working in a couple of hours, but need to iterate. Right now, you can click on the time series and get the anatomical map (which is a bit slow), but not go the other way round.

Feel free to comment.

sarangnemo commented:
cc @britta-wstnr

larsoner commented:
Cool! I would try it in a single figure with two rows, brains in the first row and time course in the bottom.

Things to potentially keep in mind:

- This eventually should work for all source spaces, not just volume (I assume what it currently is limited to?), but will require some of what @TommyClausner is working on: morphing surf/vol/mixed to other subjects and/or to a NIfTI volume.
- Eventually we probably want a mode for glass-brain plotting (like #4496) or plotting on individual MRI (like here). So consider that while coding (modularity) and choosing the API (don't make too many options specific to this style of plotting, if possible).
- One thing that sped up #4496 (should be doable since the relevant PR to nilearn was merged long ago) was adding `resampling_interpolation='nearest'` to the call to `plot_glass_brain`, perhaps something similar exists for whatever plotting function you are using. In any case, line profiling made it pretty easy to find the offending `nilearn` lines.

larsoner commented on mne/viz/_3d.py:
```diff
@@ -1717,6 +1717,161 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+
+            # the affine transformation can sometimes lead to corner
+            # cases near the edges?
+            if np.any(cut_coords < 0):
```
Maybe a little bit below zero but I doubt it should be `< -1`. In any case a `np.clip` here might be better in case some values are `>= shape[ii]`, too

larsoner commented on mne/viz/_3d.py:
```diff
@@ -1717,6 +1717,161 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+    params = {'stc': stc, 'ax_time': ax_time, 'lx': None,
+              'plot_map_callback': plot_map_callback,
+              'img_idx': index_img(img, idx),
+              'title': 'LCMV (t=%.3f s.)' % stc.times[idx],
```
`'Activation'` or just `t=` here, no guarantee this is LCMV

larsoner commented on mne/viz/tests/test_3d.py:
```diff
@@ -395,6 +399,20 @@ def test_snapshot_brain_montage():
+    n_time = 5
+    data = np.random.RandomState(0).rand(n_verts, n_time)
+    stc = VolSourceEstimate(data, vertices, 1, 1)
+    stc.plot(sample_src, subject='sample', subjects_dir=subjects_dir)
```
Add test that it currently raises error if not `VolSourceEstimate`

larsoner replied:
... for `plot_volume_source_estimates`

larsoner commented on mne/viz/_3d.py:
```diff
@@ -1717,6 +1717,161 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+                          event.ydata)
+        if event.inaxes is ax_z:
+            cut_coords = (event.xdata, event.ydata,
+                          params['ax_x'].lines[1].get_ydata()[0])
```
these conditionals should be moved inside the one in `1801`, the second should be `elif` and the last should be `else`

larsoner commented on mne/viz/_3d.py:
```diff
@@ -1717,6 +1717,161 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+        The path to the freesurfer subjects reconstructions.
+        It corresponds to Freesurfer environment variable SUBJECTS_DIR.
+    mode : str
+        The plotting mode to use. Either 'anat' (default) or 'glass_brain'.
```
Should be `stat_map` to mirror the `nilearn` name

larsoner commented:
Oh BTW I think the first selected voxel should be the global (spatio-temporal) max active one

jasmainak commented:
@larsoner this is quite close to done I would say. I implemented MIP, fixed circle/travis. I'll add a couple of `_fake_clicks` later in the day today, but I think that would be it for me. Let me know if you have any additional feature requests or if we can do a final cleanup commit, update `whats_new` and move on to `visbrain-mne` stuff :-)

sarangnemo commented:
Is it possible to visualize results on Nifti-format MRI images? (This is a common need around here.) If not, would it be straightforward to implement?

agramfort commented:
you mean with full MRI resolution?

sarangnemo commented:
@agramfort The underlying MRI should be full-resolution, while the functional image need not be resampled. e.g., we routinely plot source maps with 5mm grid resolution on a full-res Nifti-format MRI with 1mm resolution.

agramfort commented:
ok got it. I don't know. Let's see what @jasmainak says

jasmainak commented:
I'm not sure I follow -- not an MRI expert :) Is the `T1.mgz` file not good enough? If so, I can expose the [`bg_img` param of `plot_stat_map` in nilearn](http://nilearn.github.io/modules/generated/nilearn.plotting.plot_stat_map.html), so you can use whatever image you like. Would that do?

larsoner commented:
My vote is to deal with non-T1 in another PR (@jasmainak perhaps open an issue?). It might get a bit (more) complicated due to coordinate frame differences / needing to resample to the right space.

@jasmainak I'll look at the code and test it out hopefully tomorrow, if it's good to go from your end please set to MRG

jasmainak commented:
sorry for late response. I simply exposed the `bg_img` param of nilearn. I _hope_ it works out of the box, but I don't have data to test it myself. @sarangnemo feel free to give it a try. Otherwise, I'll revert it and address in another PR

agramfort commented:
If I use nilearn > 0.4 it works for me.

However:
- the ylabel is not visible
- if I click outside of the head in the black I still see a non-flat brain signal. How can this be?
- it's quite slow to update after the click. It's nilearn plotting which is the slow part?

jasmainak commented:
okay, I took care of @agramfort 's concerns. Just a couple of responses below:

> if I click outside of the head in the black I still see a non-flat brain signal. How can this be?

good question. There are two things going on:

- *Thresholding of the image*: Even if a voxel is dark, it could be because the intensity is below the threshold. Earlier this threshold was hard-coded but now it is exposed to the user.
- The images on the top subplots are shapshots at a *single time instant*: That is to say, even if a voxel is black at the current time instant, it could be colored at another time instant. To make this more obvious, now we also show the threshold on the time axis with horizontal dotted lines. This way, one can infer at what time instant the selected voxel will exceed the threshold.

> It's nilearn plotting which is the slow part?

yep, for most of the part. In the glass brain plot, I need to resample the image to compute the maximum intensity projection which is a little slow. However, we do it when the mouse moves into one of the top subplots so that it is computed before the user clicks.

agramfort commented:
the MNE source space is limited to the brain so It cannot be that clicking
outside of the brain leads to a non-zero activation.
Technically  you should be outside of the array limits.

jasmainak commented:
sure, but did you try again? It shouldn't be happening anymore

britta-wstnr commented:
Hi @jasmainak , great work!!
I played around a bit with your code to test the ``nii`` support -- and it loads the nifti without problems! However, the figure is not interactive for me. Other interactive plots, e.g., for ``raw`` work well. Is there a particular backend to use for plotting? -- I am on Linux, so far I tried Python 2 (does it require 3?). Any hints welcome!

jasmainak commented:
Also, maybe try running the example in this pull request first?

britta-wstnr commented:
Mh, I already tried the ``plot_lcmv_beamformer_volume`` and added the call to your function, also with the ``mgz`` instead of ``nii``. Is that the example you mean?

jasmainak commented:
> Mh, I already tried the plot_lcmv_beamformer_volume and added the call to your function,

hmm ... did you clone this branch instead of modifying it locally? If you already did that, I have no clue honestly :-\ Did you run it from `ipython`? What is your `matplotlib` version? Maybe you need to update it?

britta-wstnr commented:
Yes, I cloned the branch. -- I tried both, ``ipython`` and ``python`` - but I will check out whether updating ``matplotlib`` does the trick tomorrow!

britta-wstnr commented:
No luck with recent ``matplotlib`` either. Also waited some time to see whether it is just really slow - but nothing happened and no active processes either. I can try in a fresh ``conda`` environment next week, although I cannot really see which weird interactions there could be - are you using MNE-Python's ``yml``-file or would you mind sending me yours otherwise? 
Another last hope maybe: which ``matplotlib`` backend are you using?

jasmainak commented:
> No luck with recent matplotlib either. Also waited some time to see whether it is just really slow - but nothing happened and no active processes either. I can try in a fresh conda environment next week, although I cannot really see which weird interactions there could be - are you using MNE-Python's yml-file or would you mind sending me yours otherwise?

I had my own environment but I just tried with MNE-Python's yml-file and it worked just fine for me.

> Another last hope maybe: which matplotlib backend are you using?

I started ipython with 

$ipython --matplotlib

jasmainak commented:
@britta-wstnr do you see any error messages on the ipython console?

britta-wstnr commented:
Yep, it was the backend! Perfect, thanks! :smile:

jasmainak commented:
Great! So, what's left here to do?

britta-wstnr commented:
As far as I can see adding a Nifti via ``bg_img`` works fine, cc @sarangnemo
-- at least for the ``sample`` data with a ``nii`` converted from the ``mgz``.

agramfort commented:
works fine for me.

comments:
- I don't see the glass_brain mode illustrated in examples.
- I think we should have an entry in the visualization section in http://martinos.org/mne/stable/documentation.html on source estimates. Basically showing every option to visualize source estimation results : on surface, on volume, dipoles.

larsoner commented:
- This is what I see for the plot:

  <img width="912" alt="screen shot 2018-08-07 at 15 13 56" src="https://user-images.githubusercontent.com/2365790/43802978-ae36d440-9a54-11e8-9a69-ef114f926e88.png">

  Is it possible to disable the `Activation` tag or move it into the bottom plot instead of the top? It blocks the orientation labels and part of the brain.

- When in `glass_brain` mode and I click a voxel, I get this warning:

  ```
  MatplotlibDeprecationWarning: Adding an axes using the same arguments as a previous axes currently reuses the earlier instance.  In a future version, a new instance will always be created and returned.  Meanwhile, this warning can be suppressed, and the future behavior ensured, by passing a unique label to each axes instance.
  ```

- Is it possible to instead of using `threshold`, use `fmin`, `fmid`, and `fmax` like we do for surfaces? It will require some work to customize/adjust the `cmap` to these nonlinear bounds (and probably setting `threshold=fmin`), but it's probably worth it for consistency with the surface variants.

- Does it work properly for two-sided / signed data? `fmin` / `fmid` / `fmax` work differently for surface source estimates, creating symmetrical positive and negative bounds. We should probably also have this functionality here. (This will probably require `cmap` adjustment and then `threshold=-fmax` in the end.)

jasmainak commented:
Let me just add here a couple of feature requests from @sarangnemo that we discussed at Montreal. These are what the NutmegTrip viewer has currently:

- Visualization of vector fields on top of functional maps (after @britta-wstnr implements the computation aspect)
- Ability to save an animation of the functional map
- Getting a max: temporal / spatial (globally or within x mm of mouse click) / spatiotemporal

Did I forget something?

agramfort commented:
that can be done in following PRs

first address the reported issues on this one.

larsoner commented:
Setting back to WIP, feel free to set back to MRG once the comments in https://github.com/mne-tools/mne-python/pull/5329#issuecomment-411206323 are addressed. I think it's not ready to go until they are

jasmainak commented:
Sorry I haven't had time to get back to this. I'll be on it soon ...

agramfort commented:
@jasmainak you need to rebase

please put the PR back to MRG when ready.

jasmainak commented:
> When in glass_brain mode and I click a voxel, I get this warning:

@larsoner what version of `matplotlib` are you using? I could reproduce this only with 2.2.3 but then I can't develop on this because my breakpoints don't work anymore on this version ...

jasmainak commented:
> Is it possible to instead of using threshold, use fmin, fmid, and fmax like we do for surfaces? It will require some work to customize/adjust the cmap to these nonlinear bounds (and probably setting threshold=fmin), but it's probably worth it for consistency with the surface variants.

I can't seem to find the `fmin`, `fmid` format in the code base. It seems it has been deprecated. Do you mean `clim`?

jasmainak commented:
> Does it work properly for two-sided / signed data? fmin / fmid / fmax work differently for surface source estimates, creating symmetrical positive and negative bounds. We should probably also have this functionality here. (This will probably require cmap adjustment and then threshold=-fmax in the end.)

`plot_stat_map` have a parameter called `symmetric_cbar` which should adjust the limits automatically depending on the sign of the data. I'm not sure if it's a good idea to override too many of the `nilearn` defaults ...

agramfort commented:
how much work is needed to support the same API for surface plots and
volume plots?
nilearn allows you to set what?

jasmainak commented:
So, let me go through the comments one more time as it is somewhat scattered:

> I think we should have an entry in the visualization section in http://martinos.org/mne/stable/documentation.html on source estimates. Basically showing every option to visualize source estimation results : on surface, on volume, dipoles.

I'd suggest leaving this for another PR unless you insist ...

> When in glass_brain mode and I click a voxel, I get this warning:

I'm still fighting with this. I can't debug easily in this `matplotlib` version where I can reproduce this warning (2.2.3). Not sure if it's just me or something was changed in `matplotlib` but I get this warning: "QCoreApplication::exec: The event loop is already running" on using `pdb.set_trace` and `ERROR:root:No traceback has been produced, nothing to debug.` if I try to use postmortem debugging on ipython

> Is it possible to disable the Activation tag or move it into the bottom plot instead of the top? It blocks the orientation labels and part of the brain.

done

> Is it possible to instead of using threshold, use fmin, fmid, and fmax like we do for surfaces? It will require some work to customize/adjust the cmap to these nonlinear bounds (and probably setting threshold=fmin), but it's probably worth it for consistency with the surface variants.

The MNE API supports `clim` now -- so I guess I could harmonize this but most of the API is anyway different

> Does it work properly for two-sided / signed data? fmin / fmid / fmax work differently for surface source estimates, creating symmetrical positive and negative bounds. We should probably also have this functionality here. (This will probably require cmap adjustment and then threshold=-fmax in the end.)

Two-sided signed data should already work (because of the `symmetric_cbar` param in `nilearn` which is `"auto"` by default) but I haven't tested it. I could redo this to use `clim` I suppose. It would be one day of work for me, but I can't get to it until 29th since I'm traveling again ...

> how much work is needed to support the same API for surface plots and volume plots?

the API is anyway quite different. Surface plots use pysurfer as backend, so the API is pysurfer-like. Here, we are using `nilearn` so I suggest we stick to `nilearn` API.

larsoner commented:
I would rather have all MNE plotting functions use the same clim
conventions and behavior, namely that of MNE-C (we changed PySurfer to make
it support this). It's nicer for users.

jasmainak commented:
@larsoner I implemented the `pysurfer` style fmin/fmid/fmax (copying some code over from there). But somehow the colorbar looks bizarre to me. If I append the following lines to the example:

```py

import numpy as np
import matplotlib.pyplot as plt
from mne.viz.utils import _get_pysurfer_cmap

colormap = 'hot'
ctrl_pts = np.array([0.48665809, 0.55441409, 1.12674341])
center = 0.752
cmap = _get_pysurfer_cmap(colormap, ctrl_pts, center)

data = stc.data.max() * np.random.random((8, 8))
plt.figure()
plt.subplot(121)
plt.pcolormesh(data, cmap=colormap)
plt.colorbar()

plt.subplot(122)
plt.pcolormesh(data, cmap=cmap)
plt.colorbar()
plt.show()
```
I get this:

![fmin_fmid](https://user-images.githubusercontent.com/15852194/44948938-45890300-adf5-11e8-822e-bcc9cd3f026a.png)

So, it does seem that fmid is mapped to the right value but the lower side of the colorbar looks squished to me. Let me know if this looks normal to you.

larsoner commented:
You did not set the vmin and vmax in your pcolormesh call. You probably need to do it doesn't just use the min and max of the data. Also those arguments are not what I'd expect, I thought center would only be used for a two sided colormap.

But in any case it does not look correct. Plot some random data in PySurfer with one sided and two sided colormap behaviors and you should see this isn't correct. If your fmin is 0.5 and fmax 1.1, then your cbar limits should show 0.5 to 1.1 for one sided and -1.1 to 1.1 for two sided variants

larsoner commented:
(this is how STC.plot currently handles things at least, which is the API we want to conserve)

jasmainak commented:
> Also those arguments are not what I'd expect, I thought center would only be used for a two sided colormap.

yeah `center` is not used. I just passed it to make the function happy and to demonstrate the problem.

>  If your fmin is 0.5 and fmax 1.1, then your cbar limits should show 0.5 to 1.1 for one sided

True, but that's just because of the last `alpha` channel in the colormap right? You make it transparent? It doesn't seem to have any effect when I do the same ...

larsoner commented:
At least one reason your cbar limits are not set priority in this example because they will default to the data min and max, as you did not explicitly set the vmin/vmax/clim (can't remember which is used) in the pcolormesh call.

jasmainak commented:
okay so if I give the cbar `fmin` and `fmax` as the limits I get this:

![new_mapping](https://user-images.githubusercontent.com/15852194/44949321-1bd3da00-adfd-11e8-9968-e0a0b3ced1cf.png)

Now, it does look correct to me. The `fmin` and `fmid` are so close that the lower part of the colorbar is compressed. The only question is weather the control points are correct or not. I simply called this function:

```py
    ctrl_pts, colormap, _, _ = _limits_to_control_points(
        clim, stc.data, colormap, transparent=True, fmt='matplotlib')
```
and those were the values I got for the LCMV example ...

larsoner commented:
The right plot looks reasonable to me. Hopefully just minor tweaks are necessary to get it to work for two sided now. Then I think it should replicate the existing API, right?

jasmainak commented:
okay, so it seems `nilearn` expects the colormap to be symmetrical. From the documentation, it says:

> cmap: The colormap for specified image. The ccolormap must be symmetrical.

Unfortunately, it silently allows sequential colormaps but does not handle it properly. So, we should throw an error on the MNE side for this.

larsoner commented:
So you cannot find way to make the two sided data case (mne colormap, which is always symmetric about zero in our API BTW) work properly? What about the one sided case (usually hot colormap for us), does that work properly?

larsoner commented:
... If so I'm surprised the two sided case can't work, though, as it seems in a worst case you could tell nilearn that it's one sided and set the lower limit to -fmax and upper limit to +fmax and pass it a ( nonlinearly scaled) colormap that shows the -fmax,-fmid,-fmin,0,fmin,fmid,fmax points in the correct places. Under the hood Mayavi has no sense of one or two sided data so this logic already lives in PySurfer and was also ported to the GSoC project

agramfort commented:
if you make it work with  mne or mne_analyze cmaps it's good enough for a
first contrib. Then you can put a warning when different saying it's not
fully tested.

larsoner commented:
> The problem is that there isn't a straightforward way to check if a colormap is divergent 

IIRC from our end the API contract is basically that passing `pos_lims` means that colormap should be interpreted as divergent, if `lims` is passed it's interpreted as sequential. You can probably mimic both of these behaviors, `pos_lims`/divergent by doing the necessary nonlinear rescaling, and `lims`/sequential by only duplicating the colormap and only using the upper half. But I agree with @agramfort that for now if you want you can restrict the use cases to only support `pos_lims` arguments and the `mne` colormap for now, raising an error for `lims` argument.

Then someone else can do a follow-up PR to hack together the sequential / one-sided case.

At some point we might also want to keep track of the list of diverging colormaps. We have some checks for `mne` in our auto limit setting functions that should really probably be checks for any diverging colormap.

jasmainak commented:
> But I agree with @agramfort that for now if you want you can restrict the use cases to only support pos_lims arguments and the mne colormap for now, raising an error for lims argument.

okay so for now it will throw an error if it's a recognized sequential colormap from `maptlotlib`. By default it will use `'mne'` colormap. I checked and it seems to work both with one sided and two sided data.

What's left to do? Just the warnings and Travis, right?

larsoner commented:
So you can't use the hot colormap for one sided data?

jasmainak commented:
nopes, not right now. I could do the duplication trick that you mentioned to allow sequential colormaps. As you wish ...

larsoner commented:
It would be best to have the APIs be the same. It might actually be less work than trying to catch and properly raise errors

agramfort commented:
@jasmainak can you help me understand why it is complicated / a pain to do?

larsoner commented:
[Here you go](https://matplotlib.org/examples/color/colormaps_reference.html):
```
         ('Diverging', [
            'PiYG', 'PRGn', 'BrBG', 'PuOr', 'RdGy', 'RdBu',
            'RdYlBu', 'RdYlGn', 'Spectral', 'coolwarm', 'bwr', 'seismic']),
```
They do not update this list so frequently so it should be okay. We can add `seaborn` ones if you want, too.

Really users can still specify whether the `cmap` should be treated as `diverging` or `sequential` by passing in `lims` or `pos_lims`, so we are really only talking about auto-detection here, too.

jasmainak commented:
yes, we already have that in the PR though

larsoner commented:
> yes, we already have that in the PR though

What do you mean by "that"? A list of diverging colormaps? Then you should be good to go in terms of inferring what is diverging or not based on what I wrote above:

> Let's keep it simple. If it's in the mpl list for diverging maps (it's online somewhere we just need to copy it), it's diverging. Otherwise, sequential.

And from what I understand this was all that was making it difficult to implement the solution I proposed above about mirroring the `sequential` colormap to satistify nilearn's requirement/behavior that all colormaps be treated like diverging ones.

jasmainak commented:
yes, we already have the list.

> And from what I understand this was all that was making it difficult to implement the solution I proposed above about mirroring the sequential colormap to satistify nilearn's requirement/behavior that all colormaps be treated like diverging ones.

Correct, but this was not so straightforward. I got it to work now, but don't ask me how ... see the commit. Just copying the `pysurfer` code and mirroring wouldn't work ...

larsoner commented:
> I got it to work now, but don't ask me how

This does not bode well for it working long-term / being bug-free :(

> Just copying the pysurfer code and mirroring wouldn't work ...

It looks like now in our code we don't have the same thing that lives in PySurfer. That's not good for future compatibility -- now if some bug is fixed in PySurfer we can't just copy it over. And it seems like what's going on in this code already isn't so well understood, in the future it will only be less so.

Would it help to have another set of eyes look at it / take a stab at coding it?

agramfort commented:
I don't know @jasmainak what i can do to help. But this has been open for a while now. We should converge....

jasmainak commented:
> Would it help to have another set of eyes look at it / take a stab at coding it?

Sure, feel free to [simplify it](https://github.com/mne-tools/mne-python/blob/f3a90cf9a9d86a651d239c607e7ed53f4023c913/mne/viz/utils.py#L104-L111). You can push directly to this branch if you like ...

larsoner commented:
I almost have things working. FYI there is a related PR to fix `nilearn` (though I have a workaround in the code I'm making, too):

https://github.com/nilearn/nilearn/pull/1758

larsoner commented:
Okay I simplified/unified the code a bit and rebased to get rid of the merge conflict. 

Now the only divergent colormap supported is `mne`. This reduces the functionality that you had, but it comes with a big benefit. We can add proper support for other divergent colormaps in `_limits_to_control_points` and it should work properly for the volume *and* surface plotters. I don't think it would be too much work to do it. We can also ensure that `transparent=True` makes the middle bits transparent for these colormaps.

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1927,44 +1943,74 @@ def _onclick(event, params):
+        axes.clear()
+        fig.axes[-1].clear()  # cbar
+        with warnings.catch_warnings(record=True):  # nilearn bug; ax recreated
+            warnings.simplefilter('ignore', MatplotlibDeprecationWarning)
```
ha! I was trying to get around it with `delaxes` but I guess this is easier ...

larsoner replied:
Yes this is a bug that should be fixed upstream ideally. In the meantime you're right that `delaxes` might be better if it's possible, but this should probably be okay

jasmainak commented:
great @larsoner ! I think we agree more now that you actually hacked the code ;)

I agree it's better to deal with the divergent colormaps in a unified way in a separate PR ...

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1718,6 +1719,310 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+    plot_kwargs = dict(
+        threshold=None, axes=axes,
+        resampling_interpolation='nearest', vmax=vmax, figure=fig,
+        colorbar=True, bg_img=bg_img_param, cmap=colormap, black_bg=True,
```
This will ignore the `colorbar` param

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1718,6 +1719,310 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+            # Hopefully this is a sufficiently rare use case!
+            raise RuntimeError('Negative values not supported currently, '
+                               'Consider shifting or flipping the sign of '
+                               'your data for visualization purposes')
```
L1943-L1945 should be changed then?

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1718,6 +1719,310 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+    if np.array_equal(ctrl_pts, scale_pts):  # set eq above iff one-sided
+        # there is a bug in nilearn where this messes w/transparency
+        # Need to double the colormap
+        if (ctrl_pts < 0).any():
```
actually this won't be negative even if the data is signed

larsoner replied:
the user can pass negative control points, I'll clarify the message and add a test

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1718,6 +1719,310 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+            # use arbitrary bounds :(
+            # Should get them to support non-mirrored colorbars, or
+            # at least a proper `vmin` for one-sided things.
+            # Hopefully this is a sufficiently rare use case!
```
so we don't support signed data now?

larsoner replied:
Don't currently support sequential data where the lower limit is < 0

jasmainak commented:
also, I tried the viz with signed data and `clim="auto"`. It gives me weird results ...

jasmainak commented on mne/viz/tests/test_3d.py:
```diff
@@ -395,33 +395,34 @@ def test_snapshot_brain_montage():
     for mode in ['glass_brain', 'stat_map']:
-        with pytest.warns(DeprecationWarning,
-                          match='scalars to be interpreted as an index'):
+        with pytest.warns(None):  # sometimes get scalars/index warning
```
why did you change this? Wasn't it better to catch the exact warning in question?

larsoner replied:
Tests failed as it's not emitted on all platforms (this is our standard way of handling these cases)

agramfort commented:
in https://9346-1301584-gh.circle-artifacts.com/0/html/_images/sphx_glr_plot_lcmv_beamformer_volume_004.png
is it possible not to see the negative part of the colorbar?

larsoner commented:
> is it possible not to see the negative part of the colorbar?

Not directly via `nilearn`. We could probably hack it with setting the `colorbar` axis limits ourselves, though.

larsoner commented:
> also, I tried the viz with signed data and clim="auto". It gives me weird results ...

I used `clim = 'auto'` in the example and I get something reasonable, so you'll have to be more specific:

![screenshot from 2018-09-13 15-12-11](https://user-images.githubusercontent.com/2365790/45510165-6a6f6580-b767-11e8-92dd-4b8b74115cf3.png)

jasmainak commented:
I tried the following in your example:

```py
stc.plot(src=forward['src'], subject='sample', subjects_dir=subjects_dir, 
            colorbar=True, mode='glass_brain', colormap='mne_analyze')
```

and I got this:

![weird](https://user-images.githubusercontent.com/15852194/45511849-1915a500-b76c-11e8-811c-baf652e9aee0.png)

dengemann commented:
Nice!!

On Thu, Sep 13, 2018 at 9:46 PM Mainak Jas <notifications@github.com> wrote:

> I tried the following in your example:
>
> stc.plot(src=forward['src'], subject='sample', subjects_dir=subjects_dir,
>             colorbar=True, mode='glass_brain', colormap='mne_analyze')
>
> and I got this:
>
> [image: weird]
> <https://user-images.githubusercontent.com/15852194/45511849-1915a500-b76c-11e8-811c-baf652e9aee0.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/5329#issuecomment-421129864>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB0fir5W-e8JLRBwlyLeE7f40EaawFT9ks5uarYSgaJpZM4VIZht>
> .
>

jasmainak commented:
Also the colorbar goes up to 0.94 but I know the max of the data is 1.5 from the time series. So, as a user, I would be a bit surprised. Is that how the surface plots have been shown so far as well ?

larsoner commented:
> and I got this:

I suspect that the `nilearn` plotter shows the `abs` of the data, is it possible? Not sure what we should do, if anything else, in this case.

> Also the colorbar goes up to 0.94 but I know the max of the data is 1.5 from the time series. So, as a user, I would be a bit surprised. Is that how the surface plots have been shown so far as well ?

It should get the `clim` the same way the surface ones do, too, using the percentiles of the data, yes.

agramfort commented:
try this

from nilearn.plotting.img_plotting import _crop_colorbar
from nilearn import plotting
from nilearn import datasets

motor_images = datasets.fetch_neurovault_motor_task()
stat_img = motor_images.images[0]

# with glass brain
display = plotting.plot_glass_brain(stat_img, plot_abs=False, colorbar=True)
_crop_colorbar(display._cbar, 0, None)

# with stat map
display = plotting.plot_stat_map(stat_img, colorbar=True)
_crop_colorbar(display._cbar, 0, None)
plotting.show()

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1718,6 +1719,323 @@ def _get_ps_kwargs(initial_time, require='0.6'):
+    mode : str
+        The plotting mode to use. Either 'stat_map' (default) or 'glass_brain'.
+        For "glass_brain", activation absolute values are displayed
+        after being transformed to a standard MNI brain.
```
The time series still shows me signed values but the stat map shows absolute values. They should be made consistent, no?

larsoner replied:
I don't think they need to be. The glass brain does abs for the MIP so we can't and shouldn't change that, and I see no reason to distort the time series unnecessarily. These seem like doc issues to me.

jasmainak commented:
@larsoner if I do this:

```py
stc.plot(src=forward['src'], subject='sample', subjects_dir=subjects_dir,
         colorbar=True, mode='stat_map', colormap='mne')
```
I get
![mne_cmap](https://user-images.githubusercontent.com/15852194/45527386-46cd0f00-b7a9-11e8-9753-899ce74b6eb7.png)

even though there are no negative values, the colorbar has negative values.

Also, is the 'mne' and 'mne_analyze' colormap documented somewhere? I could find only
find [this](https://martinos.org/mne/stable/generated/mne.viz.mne_analyze_colormap.html) but I think the docstring of `cmap` should point to it or mention what they are and the fact that they are used when `colormap="auto"`. Are they both signed colormaps?

larsoner commented:
mne and mne_analyze are aliases

That mne behavior always being two sided is what surface maps should also do, so changing that behavior would be another PR (but probably not worth it because it would break peoples code

jasmainak commented:
okay no other comments from my end!

jasmainak commented:
> That mne behavior always being two sided is what surface maps should also do, so changing that behavior would be another PR (but probably not worth it because it would break peoples code

yes that's fine. I guess my point was that they need to be documented ...

agramfort commented:
<img width="1289" alt="screen shot 2018-09-14 at 09 55 25" src="https://user-images.githubusercontent.com/161052/45537394-67768280-b804-11e8-87b3-4d0c30e0180a.png">

2 tiny issues:
- the 0.5 xticks label is cropped
- I find the white cross on the right not visible. can we do better?

suggestion:
- be able to use arrows to move in time like we have eg in mne_analyze (right arrow : move one sample forward, shift+right arrow move by 10ms and the reverse for left arrow)

agramfort commented:
<img width="648" alt="screen shot 2018-09-14 at 15 06 34" src="https://user-images.githubusercontent.com/161052/45552023-db7a5000-b82f-11e8-8795-adb348d3e5b2.png">

now ylabel is cropped for me. Is it just me?

larsoner commented:
Updated to use `tight_layout` now so there shouldn't be clipping problems anymore

larsoner commented:
All better:

https://9367-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_lcmv_beamformer_volume.html

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1987,22 +1984,22 @@ def _onclick(event, params):
+            params['fig_anat'] = partial(
+                plot_func, **plot_kwargs)(*args, **kwargs)
+        for key in 'xyz':
+            params.update({'ax_' + key: params['fig_anat'].axes[key].ax})
```
heh, neat. I was wondering how to do this ...

jasmainak commented:
oops, didn't realize we merged this. Thanks for the reviews! Yes, let's iterate more on master.
